serve/mlc_serve/engine/engine_common.py

```diff
@@ -206,25 +169,75 @@ def check_stopping_sequences(stopping_criteria, output_text, delta, is_ended):
     return output_text, delta, is_ended
 
 
-def update_sequence(
+def prepare_logprob(
+    logprob_info: Optional[List[Optional[RawLogprobsInfo]]],
+    delta: str,
+    gen_seq: GenerationSequence,
+    prompt_token_ids: List[int],
+    tokenizer: TokenizerP,
+) -> List[Optional[LogprobsContent]]:
```

Comment:
@sunggg re: above comment. It feels like we should clarify all these types/names as follow up. I feel like its not super clear what the types are. In general types like `List[Optional[Thing]]` aren't very clear. Especially given we multiple of them appearing like `Optional[List[Optional[RawLogprobsInfo]]]` in the same function signature.

Reply:
Agreed. Added this comment to the ticket to follow-up.

Reply:
I think I had to do it to make mypy happy. Hope we can follow-up since I also want to improve this messy logprob-related types in general.